src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs

```diff
@@ -0,0 +1,333 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Reflection;
+using System.Reflection.PortableExecutable;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+internal enum Compatibility
+{
+    Compatible,
+    Incompatible,
+    Inconclusive
+}
+
+internal sealed class InconclusiveCompatibilityCollection
+{
+    private Dictionary<string, HashSet<string>> _data = new();
+
+    public bool IsEmpty => _data.Count == 0;
+
+    public void Add(string assyName, string namespaceName, string typeName)
+    {
+        HashSet<string>? incAssyTypes;
+
+        if(!_data.TryGetValue(assyName, out incAssyTypes))
+        {
+            incAssyTypes = new();
+            _data.Add(assyName, incAssyTypes);
+        }
+
+        incAssyTypes.Add(string.Join(":", namespaceName, typeName));
+    }
+
+    public HashSet<string> EnumerateForAssembly(string assyName)
+    {
+        if(_data.TryGetValue(assyName, out HashSet<string>? incAssyTypes))
+            return incAssyTypes!;
+
+        return new HashSet<string>();
+    }
+}
+
+internal sealed class MinimalMarshalingTypeCompatibilityProvider : ISignatureTypeProvider<Compatibility, object>
```

Comment:
I wonder if it's worthwhile importing the type system from NativeAOT (just add the sources) and actually constructing a full type universe.  (in https://github.com/dotnet/runtime/tree/main/src/coreclr/tools/Common/TypeSystem - see for example how the PGO tool does it for the TypeRefTypeSystemContext and the TraceTypeSystemContext https://github.com/dotnet/runtime/tree/main/src/coreclr/tools/dotnet-pgo

In particular I wonder if we can reuse their typesystem's definition of blittability rather than inventing our own

Reply:
Well, their `TypeDesc` certainly handles much easier than `S.R.Metadata`. On the other hand, we have a working solution now (that possibly needs some more tweaking). Can we use it for the time being and move to NAOT type system as a follow up?